### PR TITLE
Quota UI

### PR DESCRIPTION
**Pull request by utchoang**
Open for https://github.com/apache/cloudstack-primate/issues/295
@rhtyd cc @svenvogel 
* [x] Quota Summary: Selected account for open the account quota
* [x] Quota Summary: Add button for add credits
* [x] Quota Summary: Show Quota Usage and Balance in detail
* [ ] Quota Tariff: Missing calendar functionality
* [x] Quota Tariff: change the tariff value
* [x] Quota Email template: Search is not working

### CODE REVIEW

**Comment by utchoang** [comment]
* Selected account quota
image

**Comment by utchoang** [comment]
* Button Add Credits
image

* Form Add Credits
image

**Comment by utchoang** [comment]
* Quota Usage
image

**Comment by utchoang** [comment]
* Quota Balance
image

**Comment by utchoang** [comment]
* Quota Tariff: change the tariff value
image
image
image
image

**Comment by utchoang** [comment]
* Search Quota Email template

image

**Comment by utchoang** [comment]
Hi @rhtyd . Regarding Issue https://github.com/apache/cloudstack-primate/issues/210 (Missing calendar functionality in Quota > Tariff), I found there is a conflict as I restructured the search function in PR https://github.com/apache/cloudstack-primate/pull/235. Is it possible for this issue to be solved after PR https://github.com/apache/cloudstack-primate/pull/235 is merged?

**Comment by rhtyd** [comment]
Hi @utchoang can you fix the conflicts? To answer your question - yes if it makes it easier we can address any specific case after merging #235.
Is this PR otherwise ready for review/testing and does it address all of:
#209
#210
#211
#212

Thanks.

**Comment by utchoang** [comment]
@rhtyd cc @svenvogel I completed resolved conflict. Please review it. Thanks

**Comment by rhtyd** [comment]
@utchoang can you fix the conflict? Thanks.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

File: src/locales/en.json [hunk 1]
```diff
@@ -2384,6 +2388,148 @@
 "message.error.number": "Please enter a valid number",
 "message.error.required.input": "Please enter input",
 "message.error.select": "Please select option",
+<<<<<<< HEAD
```
Comment (rhtyd) [inline src/locales/en.json:2391]: Merge issues @utchoang

**Comment by rhtyd** [comment]
@utchoang I'm trying to review and merge old PRs, sorry for participating late; can you address the following:
- Fix merge conflicts, see issues of merged conflicts changes in translation file
- Make minimum changes to resourceview and listview if possible
- Revisit and refactor this PR, with goal to simplify code wherever possible

Thanks.

**Comment by rhtyd** [comment]
@utchoang the PR title says WIP, is it ready for review and testing?

**Comment by utchoang** [comment]
@rhtyd Yes. Please review it. Thanks.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@utchoang LGTM but can you address the following:
- Fix translation/key label usage across views (see locales/en.json to pick the keys, all lowercase)
- Add ability to edit the email template (some action/button to edit the email template), refer legacy UI for implementation (but avoid changing code ListView and AutogenView)

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@utchoang there are still issues:
- Translation issue in form, for example the add credit form
- The tariff is no longer editable; if you want you can create a new view component (for ex. the search bar is not necarry for tariffs tab)

Rest LGTM; thanks.

**Comment by utchoang** [comment]
@rhtyd I've finished fixing. Please review it again. Thanks.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@utchoang can you fix the confilct please, also check all your existing PRs for conflicts; thanks

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

File: src/components/view/ListView.vue [hunk 2]
```diff
@@ -16,220 +16,272 @@
...
+      :visible="showTariffAction"
+      :closable="true"
+      :title="$t('label.quota.configuration')"
+      :confirmLoading="actionLoading"
+      style="top: 20px;"
+      @ok="submitTariff"
+      @cancel="() => { this.showTariffAction = false }"
+      centered
+    >
+      <a-form
```
Comment (rhtyd) [inline src/components/view/ListView.vue:257]: @utchoang can you undo changes in this component; instead any additional things could to pushed into it using slots (see other examples using `<slot ...`

**Comment by rhtyd** [comment]
@utchoang please avoid making changes to core widgets including ListView, ResourceView etc; as/when necessary put slots in them and create components that embed them and extend them by using slots. Happy to merge once these are addressed, thanks.

**Comment by rhtyd** [comment]
@utchoang headsup looks like there is a conflict as well

**Comment by utchoang** [comment]
@rhtyd I have made some changes to avoid changing ListView. Please review it again. Thanks.

**Comment by rhtyd** [comment]
Thanks @utchoang I'll review tomorrow
@blueorangutan package

File: src/components/view/ListView.vue [hunk 3]
```diff
@@ -141,11 +144,14 @@
       <router-link :to="{ path: '/pod/' + record.podid }">{{ text }}</router-link>
     </a>
     <a slot="account" slot-scope="text, record" href="javascript:;">
-      <router-link :to="{ path: '/account/' + record.accountid }" v-if="record.accountid">{{ text }}</router-link>
+      <router-link
+        :to="{ path: `${$route.path}/${record.account}`, query: { account: record.account, domainid: record.domainid, quota: true } }"
+        v-if="record.accountid && 'quota' in record">{{ text }}</router-link>
```
Comment (rhtyd) [inline src/components/view/ListView.vue:149]: Instead the check must be based on $route.name or path

File: src/components/view/ResourceView.vue [hunk 4]
```diff
@@ -129,6 +148,26 @@ export default {
       } else {
         return true
       }
+    },
+    fetchResourceQuota () {
```
Comment (rhtyd) [inline src/components/view/ResourceView.vue:152]: @utchoang is there a way to not make changes in core components? Here's how you can extend the, for example;
- Create required slot in AutogenView etc
- Create a new custom component for rendering the view for a router name
- In the new/separate component you can re-use AutogenView and pass any specific component in the slots for example:
```
<AutogenView ...pass any params here.... >

<span slot="actions">
... define extension here, and methods relatd to that...
</span>

</AutogenView>
```

If you follow this pattern you won't need to just extend thing inside the core components of resourceview, autogenview, listview etc; but use slot to extend them.

**Comment by rhtyd** [comment]
@utchoang thanks much better LGTM, I've left one additional remark on how you can refactor the code by use of slots (or named-slots) see - https://vuejs.org/v2/guide/components-slots.html
If you've questions let me know, I think by using slots you can avoid making any changes and re-use AutogenView in new components without duplicating logic/code for handling rest of the elements (like actions, breadcrumb etc)

File: src/components/view/ListView.vue (hunk 3 above)
Comment (utchoang) [reply src/components/view/ListView.vue]: I've changed it to check $route.path

File: src/components/view/ResourceView.vue (hunk 4 above)
Comment (utchoang) [reply src/components/view/ResourceView.vue]: I've added a new component (QuotaSummaryResource) for view summary details and undo changed ResourceView.

**Comment by utchoang** [comment]
@rhtyd Did you mean that when the zone deployment, all NSPs enabled are false? Because I understand that when the launch zone, it will enable all NSP. If I wrong, I'll undo the change this.

**Comment by utchoang** [comment]
@rhtyd I've resolved it. Please review again.Thanks.

**Comment by rhtyd** [comment]
Thanks @utchoang 
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@utchoang can you fix the conflicts? Thanks.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@utchoang I still see no usage of slots, are you working on doing a separate component implementation and use slots instead of extending AutogenView?

**Comment by utchoang** [comment]
@rhtyd I created two new components for the edit tariff value and the quota summary detail. I don't understand if I don't extend AutogenView, how can I use slots to it?

**Comment by rhtyd** [comment]
@utchoang you can create a new component and in that import and use autogenview instead of the other way around, then in your new component you can pass templates etc into slots that you define in Autogenview.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
Named slots could have been used, but I'll merge for now

**Pull request merged**